### PR TITLE
changed PID handling for section handler PIDs

### PR DESCRIPTION
**Pull request by REELcoder**
Just an initial try. The list of PIDs is now stable. I still have to check if cReceivers will get all neede PIDs delivered.

### CODE REVIEW

File: device.h [hunk 1]
```diff
@@ -40,6 +40,7 @@ class cSatipDevice : public cDevice, public cSatipPidStatistics, public cSatipBu
   cSatipSectionFilterHandler *pSectionFilterHandlerM;
   cTimeMs createdM;
   cMutex mutexM;
+  cSatipPid pidsM;
```
Comment (rofafor) [inline device.h]: This is a wrong place. You should refactor the pid handling in tuner.*

File: tuner.c [hunk 2]
```diff
@@ -405,19 +405,21 @@ bool cSatipTuner::SetSource(cSatipServer *serverP, const int transponderP, const
   return true;
 }
 
-bool cSatipTuner::SetPid(int pidP, int typeP, bool onP)
+bool cSatipTuner::SetPid(int pidP, int typeP, bool onP, bool fixP)
```
Comment (rofafor) [inline tuner.c]: s/fixP/fixedP

File: tuner.c [hunk 3]
```diff
@@ -405,19 +405,21 @@ bool cSatipTuner::SetSource(cSatipServer *serverP, const int transponderP, const
...
 {
   debug16("%s (%d, %d, %d) [device %d]", __PRETTY_FUNCTION__, pidP, typeP, onP, deviceIdM);
   cMutexLock MutexLock(&mutexM);
   if (onP) {
      pidsM.AddPid(pidP);
      addPidsM.AddPid(pidP);
      delPidsM.RemovePid(pidP);
+     if (fixP) fixedPidsM.AddPid(pidP);
```
Comment (rofafor) [inline tuner.c]: Split into two lines

**Comment by rofafor** [comment]
fixedPidsM should be cleared at least in Disconnect() (similar to add/delPidsM).

**Comment by rofafor** [comment]
Please, rebase all refactoring commits into the original one.

File: tuner.c [hunk 4]
```diff
@@ -405,19 +405,22 @@ bool cSatipTuner::SetSource(cSatipServer *serverP, const int transponderP, const
...
+bool cSatipTuner::SetPid(int pidP, int typeP, bool onP, bool fixedP)
 {
   debug16("%s (%d, %d, %d) [device %d]", __PRETTY_FUNCTION__, pidP, typeP, onP, deviceIdM);
   cMutexLock MutexLock(&mutexM);
   if (onP) {
      pidsM.AddPid(pidP);
      addPidsM.AddPid(pidP);
      delPidsM.RemovePid(pidP);
+     if (fixedP)
+       fixedPidsM.AddPid(pidP);
```
Comment (rofafor) [inline tuner.c]: Indentation doesn't match up the used VDR style (3 spaces for 'if' blocks).

**Comment by rofafor** [comment]
Add a proper description of PR. Why this is needed, how this fixes the problem, and how to validate the functionality.

**Comment by REELcoder** [comment]
OK, will do so. Had some problems to understand the git rebase topic and messe up my local repo... PR will be updated once code is tested.

**Comment by REELcoder** [comment]
I do not get the implementation in tuner.c to work. The fixedPidsM is not from the same instance when called from cSatipDevice::SetPid or cSatipDevice::OpenFilter/CloseFilter. This means CloseFilter will still cause delpids even if it has been asserted by cSatipDevice::SetPid.
I propose to go back to the implementation in device.c because there we stay in one single instance per received channel.

**Comment by rofafor** [comment]
Implemented this a bit differently - hopefully it works too.

**Comment by rofafor** [comment]
Could you explain a bit more about the mentioned OctopusNet addpid bug?

**Comment by REELcoder** [comment]
At some point I discovered that the OctopusNet does not deliver the PMT PID (164) on one channel I used frequently for testing. RTSP request sequence was the following:
```
OPTIONS rtsp://10.76.1.173/ RTSP/1.0
SETUP rtsp://10.76.1.173/?freq=186&msys=dvbc&mtype=256qam&sr=6900 RTSP/1.0
PLAY rtsp://10.76.1.173/stream=1?&pids=0,144,155,156 RTSP/1.0
DESCRIBE rtsp://10.76.1.173/stream=1 RTSP/1.0
PLAY rtsp://10.76.1.173/stream=1?freq=186&msys=dvbc&mtype=256qam&sr=6900 RTSP/1.0
PLAY rtsp://10.76.1.173/stream=1?pids=1,144,153,155,156,163,173 RTSP/1.0
PLAY rtsp://10.76.1.173/stream=1?addpids=1,128,129,130,153,163,164,173 RTSP/1.0
TEARDOWN rtsp://10.76.1.173/stream=1 RTSP/1.0
```
The PID 164 is requested by the last addpids request, but is never delivered. I wrote a small client script to modify the request sequence. After some trials it became clear that the PID 164 is not delivered if PID 163 is in the same addpids request. 
I reported this to DD and got a confirmation that they fixed the bug and a new firmware will be released soon. This was on the 17.11. Up to now this seems to have not happened, since no update is offered in the OctopusNet web interface. 

Thanks for the link to the DigitalDevices repo. I was not aware of it. Maybe https://github.com/DigitalDevices/octonet/commit/a0e2ee78dc5ee4fb32cb4a4fa907aea12282cb73 is the promised fix, I have not tried to understand this code there.

I temporarily made the workaround in tuner.c (https://github.com/rofafor/vdr-plugin-satip/pull/32/commits/9c1dc3d6ac0c0ba66b44ad0311e7b7e8b8361ac3) to avoid addpids requests. Now the PMT PID 164 is reliably delivered. 

I will test 581ac4966db8bb6badb6bd3a12ece458663ba9f5 in the next days. Thanks for this.

**Comment by rofafor** [comment]
You should enable beta firmware in your OctopusNet - I've been using those (1.0.79 at the moment) from the day one in my setup and they've bee working much nicer than than the latest stable with vdr.

**Comment by rofafor** [comment]
Any news? Can we close this one?

**Comment by REELcoder** [comment]
Hi Rolf, it works smoothly both using octonet 1.0.79 and current minisatip without any quirks enabled. I get now PMTs and other DVB-SI immediately and fully reliable when streaming using the streamdev-server plugin. Therefore I propose to close here. Thanks for your support.

**Pull request closed without merging**